### PR TITLE
docs: Add CONTRIBUTING.md, fix #1118

### LINKED ISSUE #1118: Create CONTRIBUTING.md with info for new developers
- #212 - `cargo insta`

### PR DESCRIPTION
**Pull request by MingweiSamuel**
#1118

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
Some parts are a bit sparse, and I've forgotten some things that @rohitkulshreshtha has run into getting started, I'm sure

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Contributing to Hydroflow
+
+Thanks for your interesting in contributing to Hydroflow! This is an experimental, reserach-driven
```
Comment (rohitkulshreshtha) [inline CONTRIBUTING.md:3]: Minor typo.

`s/reserach/research/`

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Commit messages should follow the [Conventional Commits specification](https://www.conventionalcommits.org/)
+when submitting a PR. Alternatively, GitHub's "Squash and merge" feature can be used to adjust the
+commit message, however those messages are not checked by CI.
+
+Commit messages are used to generate changelogs. See [Releasing](#releasing) for more.
+
+## Pull Requests
+
+CI runs a comprehensive set of tests on PRs before they are merged. This includes format and lint
+checks. To run (most of) the tests locally, you can run `./precheck.bash`. Note that this will
```
Comment (rohitkulshreshtha) [inline CONTRIBUTING.md:56]: Think precheck doesn't do WASM, correct? Any objections to adding the following to `precheck.bash`?

```shell
CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER=wasm-bindgen-test-runner cargo test -p hydroflow --target wasm32-unknown-unknown
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+overwrite any changed snapshot tests instead of failing-- you should double-check that the snapshot
+diff matches what you expect.
+
+## Snapshot Testing
+
+Hydroflow uses two types of snapshot testing: [`insta`](https://insta.rs/) and [`trybuild`](https://github.com/dtolnay/trybuild).
+Insta provides general snapshot testing in Rust, and in Hydroflow we mainly use it to test the
+graph visualization formatting of Hydroflow graphs; see `hydroflow/tests/snapshots`. `trybuild`
```
Comment (rohitkulshreshtha) [inline CONTRIBUTING.md:64]: Is testing graph visualization the main goal though? I remember you mentioning some issues that looking at the snapshot diffs can uncover. Those are useful ropes for a developer to master and are worth mentioning here.

File: CONTRIBUTING.md (hunk 1 above)
Comment (rohitkulshreshtha) [reply CONTRIBUTING.md:3]: `s/interesting/interest`

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+project which can make getting started a bit tricky. This guide will explain the project structure,
+code style, commit messages, testing setups, and more to help you get started.
+
+## Repository Structure
+
+The Hydroflow repo is set up as a monorepo and [Cargo workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).
+Relative to the repository root:
+
+* `hydroflow` is the main Hydroflow package, containing the Hydroflow runtime. It re-exports the
+  surface syntax macros which are in `hydroflow_macro` and `hydroflow_lang`. The runtime is the
```
Comment (rohitkulshreshtha) [inline CONTRIBUTING.md:13]: Nit: `s/which are//`

The sentence still makes sense without "which are".

**Comment by rohitkulshreshtha** [comment]
> Some parts are a bit sparse, and I've forgotten some things that @rohitkulshreshtha has run into getting started, I'm sure

I checked slack. Only `precheck.bash` shows up in the search. I guess that covers it!

**Comment by rohitkulshreshtha** [comment]
This looks great! Thanks for raising the bar.

File: CONTRIBUTING.md (hunk 1 above)
Comment (MingweiSamuel) [reply CONTRIBUTING.md:3]: Hmm I should probably enable spell check 😅

**Review by rohitkulshreshtha — APPROVED** [review]

**Pull request merged**